File: FCdefinition.py
from openai import OpenAI
import json
import os
import sys
import requests
import time


client = OpenAI(
    api_key="sk-wosxiisuzqcpwbnmaobpgflmgxzpumvxsuvusoduscvhcdoc",
    base_url="https://api.siliconflow.cn"
)


print("OpenAi client initialized.")

def get_definition(word:str):
    print(f"Getting definition for {word}")
    url = "https://urban-dictionary7.p.rapidapi.com/v0/define"
    
    querystring = {"term": word}
    
    headers = {
        "x-rapidapi-key": "a9f051cf53msh70504d662074c20p10960fjsn413232e29f01",
        "x-rapidapi-host": "urban-dictionary7.p.rapidapi.com"
    }
    
    # Make the request exactly like in def_test.py
    response = requests.get(url, headers=headers, params=querystring)
    
    # Process the response
    if response.status_code == 200:
        data = response.json()
        definition = data['list'][0]['definition']
        # breakpoint()
        return f"The definition of {word} is {definition}"
    else:
        print(f"Error: HTTP {response.status_code} - {response.text}")
        return f"Error: HTTP {response.status_code}"

tools=[{
    'type':'function',
    'function': {
        'name':'get_definition',
        'description':'Get the definition of a word',
        'parameters':{
            'type':'object',
            'properties':{
                'word':{
                    'type':'string',
                    'description':'The word to get the definition of'
                }
            },
            'required':['word']
        }
    }
}]

print("Making API call...")
def function_call_playground(prompt:str):
    messages=[
        {
        "role":"system",
         "content":prompt
         }
    ]
    response=client.chat.completions.create(
         model="THUDM/glm-4-9b-chat",
        messages=messages,
        temperature=0.01,
        stream=False,
        top_p=0.95,
        tools=tools
    )

    func1_name=response.choices[0].message.tool_calls[0].function.name
    func1_args=response.choices[0].message.tool_calls[0].function.arguments
    func1_out=eval(f'{func1_name}(**{func1_args})')

    messages.append(response.choices[0].message)
    messages.append({
        'role':'tool',
        'content':f'{func1_out}',
        'tool_call_id':response.choices[0].message.tool_calls[0].id
    })

    response=client.chat.completions.create(
          model="THUDM/glm-4-9b-chat",
        messages=messages,
        temperature=0.01,
        stream=False,
        top_p=0.95,
        tools=tools
    )
    # breakpoint()
    return response.choices[0].message.content

# Fix the prompt - it's missing a closing quote
prompt="what is the definition of the word 'nice'"

import time
start_time=time.time()
print(function_call_playground(prompt))
end_time=time.time()
print(f"time taken:{end_time-start_time}seconds")

File: fc-nodejs/FCdefinition.js
import {OpenAI} from "openai"
import {config} from "dotenv"
import {BASE_URL, OPENAI_API_KEY, MODEL_NAME, URBAN_DICTIONARY_API_KEY} from "./config.js"

config();

const openai = new OpenAI({
    apiKey: OPENAI_API_KEY,
    baseURL: BASE_URL
})

const getDefinition = async(word) => {
    try {

        const response = await fetch(`https://urban-dictionary7.p.rapidapi.com/v0/define?term=${encodeURIComponent(word)}`, {
            method: 'GET',
            headers: {
                'x-rapidapi-key': URBAN_DICTIONARY_API_KEY,
                'x-rapidapi-host': 'urban-dictionary7.p.rapidapi.com'
            }
        });

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const data = await response.json();
        if (data.list && data.list.length > 0) {
            console.log("data.list", data.list);
            return `The definition of "${word}" is: ${data.list[0].definition}`;
        } else {
            return `No definition found for "${word}"`;
        }
    } catch (error) {
        console.error("Error fetching definition:", error);
        return `Failed to fetch definition for "${word}": ${error.message}`;
    }
}

const tools = [{
    type: "function",
    function: {
        name: "getDefinition",
        description: "Get the definition of a word from Urban Dictionary",
        parameters: {
            type: "object",
            properties: {
                word: {
                    type: "string",
                    description: "The word to get the definition for"
                }
            },
            required: ["word"],
            additionalProperties: false
        }
    }
}];

export const function_call_playground_definition = async (prompt) => {
    const messages = [
        {
            role: "system",
            content: prompt
        }
    ];

    try {
        const response = await openai.chat.completions.create({
            model: MODEL_NAME,
            messages: messages,
            temperature: 0.01,
            stream: false,
            top_p: 0.95,
            tools: tools
        });

        const func1_name = response.choices[0].message.tool_calls[0].function.name;
        const func1_args = JSON.parse(response.choices[0].message.tool_calls[0].function.arguments);
        const func1_out = await getDefinition(func1_args.word);
        
        messages.push(response.choices[0].message);
        messages.push({
            role: "tool",
            content: func1_out,
            tool_call_id: response.choices[0].message.tool_calls[0].id
        });

        const response2 = await openai.chat.completions.create({
            model: MODEL_NAME,
            messages: messages,
            temperature: 0.01,
            stream: false
        });

        return response2.choices[0].message.content;
    } catch (error) {
        console.error("Error in function_call_playground:", error);
        if (error.response) {
            console.error("API Response:", error.response.data);
        }
        throw error;
    }
};

const prompt = "What is the definition of the word 'sabbatical'?";

const start_time = Date.now();
try {
    const result = await function_call_playground_definition(prompt);
    const end_time = Date.now();
    console.log(`Time taken for function call playground: ${(end_time-start_time)/1000} seconds`);
    console.log(result);
} catch (error) {
    console.error("Error running the program:", error);
}